src/mem_table.rs

```diff
@@ -123,75 +122,51 @@ pub(crate) struct ImmutableWal {
     table: Arc<KVTable>,
 }
 
-pub(crate) struct MemTableIterator<'a, T: RangeBounds<KVTableInternalKey>> {
+#[self_referencing]
+pub(crate) struct MemTableIteratorIner<T: RangeBounds<KVTableInternalKey>> {
+    map: Arc<SkipMap<KVTableInternalKey, RowEntry>>,
     /// `inner` is the Iterator impl of SkipMap, which is the underlying data structure of MemTable.
-    inner: Range<'a, KVTableInternalKey, T, KVTableInternalKey, RowEntry>,
+    #[borrows(map)]
+    #[not_covariant]
+    inner: Range<'this, KVTableInternalKey, T, KVTableInternalKey, RowEntry>,
```

Comment:
i am not familiar with `ouroboros`, can you explain a bit about how it works in the pull request summary?

Reply:
🤯 I didn't know either. Apparently it adds a `'this` lifetime to the struct, which fields can use. So we no longer need `'a`... seems quite nice.